### PR TITLE
client: add Nomad service registration implementation.

### PR DESCRIPTION
**Pull request by jrasell**
related https://github.com/hashicorp/team-nomad/issues/266

### CODE REVIEW

**Review by schmichael — APPROVED** [review]

File: client/serviceregistration/nomad/nomad.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+			Region:    s.cfg.Region,
+			AuthToken: s.cfg.NodeSecret,
+		},
+	}
+
+	var resp structs.ServiceRegistrationUpsertResponse
+
+	return s.retryFunc(func() error {
+		return s.cfg.RPCFn(structs.ServiceRegistrationUpsertRPCMethod, &args, &resp)
+	}, structs.ServiceRegistrationUpsertRPCMethod)
```
[the remark this answers is not in the document]
Comment (schmichael) [reply client/serviceregistration/nomad/nomad.go]: Aw phooey it's **not** safe to block. Here's an example where it's called inside a mutex: https://github.com/hashicorp/nomad/blob/5d63a133241cf2782210db8feef06f6240aac2dc/client/allocrunner/groupservice_hook.go#L111-L118

You can see here how the original Consul implementation explicitly avoids blocking in the call: https://github.com/hashicorp/nomad/blob/5d63a133241cf2782210db8feef06f6240aac2dc/command/agent/consul/service_client.go#L1106

We probably should have documented that requirement in the interface, sorry.

[the remark this answers is not in the document]
Comment (jrasell) [reply client/serviceregistration/nomad/nomad.go]: OK this is starting to become a bit clearer.

I have removed the retry functionality for now. The client RPC function itself has a very minimal retry based on any returned errors so there is some coverage there. In the event the RPC fails on register, the task will fail which seems an easy to reason initial approach with minimal opportunities for bugs. The chance of the RPC failing is small, as the client did just get the allocation from the server, but it is possible.

In the future, if we want, we could look at extracting common functionality from the Consul implementation for enqueuing and deduplicating operations. If we don't want to go down this approach, at least we would have more time to be able to write and test the code.

Hope this seems reasonable and makes sense.

File: client/serviceregistration/nomad/nomad.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+// RemoveWorkload iterates the services and removes them from the service
+// registration state.
+//
+// This function works regardless of whether the client has this feature
+// enabled. This covers situations where the feature is disabled, yet still has
+// allocations which, when stopped need their registrations removed.
+func (s *ServiceRegistrationHandler) RemoveWorkload(workload *serviceregistration.WorkloadServices) {
+	for _, serviceSpec := range workload.Services {
+		go s.removeWorkload(workload, serviceSpec)
```
[the remark this answers is not in the document]
Comment (jrasell) [reply client/serviceregistration/nomad/nomad.go]: Yes 100% agree on this and the batch approach was something I initially wanted. I couldn't, however, figure a way to handle partial delete failures.

The scenario in my head here would be that one out of several service registrations within the batch request was manually deleted by an operator earlier (by accident). The state delete for that registration would therefore fail, resulting in the transaction being aborted and the RPC returning an error. I do have an idea to populate the RPC responses with errors, and keep the generic RPC error response for unrecoverable errors. This is pretty new ground though and I thought it out of scope for this work.

This batch feature is certainly something we can add as a follow up though; hopefully I am not punting too much on work here.

File: client/serviceregistration/nomad/nomad.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (schmichael) [reply client/serviceregistration/nomad/nomad.go]: Yeah, I think this is a great place to start.

File: client/serviceregistration/nomad/nomad.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (schmichael) [reply client/serviceregistration/nomad/nomad.go]: Sounds good. I don't think we need to worry about batching or the edge case today. :+1:

**Pull request merged**